### ISSUE TITLE
Reduce marginal proof cost by eliminating queries

### ISSUE THREAD

**Issue by str4d** [zcash/orchard#168]
Currently, we query all 10 advice columns at three rotations (`-1, 0, 1`). This corresponds to 30 evaluations per circuit instance, or 960 bytes per Action. If we can adjust the gates to remove some of these rotations (either by changing the shapes of their regions, or changing where the gates are positioned / activated within the regions), we could reduce the size of the proofs.

**Comment by str4d** [comment]
I added some dev tooling in https://github.com/zcash/halo2/pull/355 to print per-constraint queries to a CSV file. From that, I was able to ascertain that the Action circuit (as of #101 merging) has several potential opportunities here:

- `advices[0]` is queried at `prev` in 4 constraints:
  - LSB check.
  - Commit<sup>ivk</sup> canonicity check
  - NoteCommit decomposition check, `b_check` (x2)

- `advices[1]` is queried at `prev` in 6 constraints:
  - LSB check
  - Decompose scalar for complete bits of variable-base mul/`y_switch`
  - Commit<sup>ivk</sup> canonicity check (x2)
  - NoteCommit decomposition/`d_check` (x2)

- `advices[2]` is queried at `prev` in 3 constraints:
  - Commit<sup>ivk</sup> canonicity check
  - NoteCommit decomposition/`e_check` (x2)

- `advices[3]` is queried at `prev` in 5 constraints:
  - Commit<sup>ivk</sup> canonicity check (x3)
  - NoteCommit decomposition/`g_check` (x2)

Just eliminating those would save 576 bytes per Action.

**Comment by str4d** [comment]
Going through the remaining advice columns' quering of `prev` for completeness (most of these won't be eliminated; if anything we might be moving queries into them):

- `advices[4]` (8):
  - Commit<sup>ivk</sup> canonicity check (x6)
  - NoteCommit decomposition check/`h_check` (x2)
- `advices[5]` (7):
  - Commit<sup>ivk</sup> canonicity check (x3)
  - NoteCommit decomposition check/`bool_check d_1` (x2)
  - NoteCommit decomposition check/`d_check` (x2)
- `advices[6]` (10):
  - q_mul_2 == 1 checks/bool_check
  - q_mul_2 == 1 checks/gradient_1
  - q_mul_3 == 1 checks/bool_check
  - q_mul_3 == 1 checks/gradient_1
  - overflow checks
  - Canonicity checks/alpha_0_prime check
  - pad-and-add
  - CommitIvk canonicity check
  - NoteCommit decomposition check/pkd_x_check (x2)
- `advices[7]` (12):
  - overflow checks (x4)
  - pad-and-add
  - CommitIvk canonicity check
  - NoteCommit decomposition check/`b3_c_prime_check` (x2)
  - NoteCommit decomposition check/`b_check` (x2)
  - NoteCommit decomposition check/`pkd_x_check` (x2)
- `advices[8]` (7):
  - Canonicity checks/`MSB = 1 => alpha_0_hi_120 = 0`
  - Canonicity checks/`z_84_alpha_check`
  - Canonicity checks/`alpha_0_prime check`
  - pad-and-add
  - CommitIvk canonicity check
  - NoteCommit decomposition check/`a_prime_check` (x2)
- `advices[9]` (14):
  - Short lookup bitshift
  - LSB check (x3)
  - q_mul_2 == 1 checks/`bool_check`
  - q_mul_2 == 1 checks/`gradient_1`
  - q_mul_3 == 1 checks/`bool_check`
  - q_mul_3 == 1 checks/`gradient_1`
  - Decompose scalar for complete bits of variable-base mul/`bool_check`
  - Decompose scalar for complete bits of variable-base mul/`y_switch`
  - NoteCommit decomposition check/`bool_check b_2` (x2)
  - NoteCommit decomposition check/`b_check` (x2)

**Comment by str4d** [comment]
Of the `cur` and `next` rotations, only one is currently in single digits:

- `advices[1]` at `next` (9):
  - q_mul_2 == 1 checks/`y_p_check` (x2)
  - Sinsemilla gate/`y check`
  - Decomposition check/`left_check`
  - Decomposition check/`b1_b2_check`
  - NoteCommit decomposition check/`e1_f_prime_check` (x2)
  - Canonicity checks/`g_0 = 1 => z13_f` (x2)

**Comment by str4d** [comment]
Commit<sup>ivk</sup> is a two-row region that is accessing `prev` and `cur`, so by altering the row its selector is enabled on, we can trivially have it query `cur` and `next`, removing all of its contributions from the above!

"LSB check" could be similarly eliminated by enabling `q_mul_lsb` one row earlier.

After these changes, the first six advice columns are now using `prev` in `[2, 3, 2, 2, 2, 4]` constraints.

**Comment by str4d** [comment]
I've examined the NoteCommit region in detail, and I don't think we can reorder any of the cells to remove `prev` queries (without introducing more in the same columns). I think this would require a second region, so we should think about how to most effectively split up the checks between two regions.

**Comment by str4d** [comment]
I refactored the NoteCommit region into multiple two-row regions, to remove all queries of `prev`. With that, we've removed 5 advice queries from the marginal proof cost, in exchange for two additional combined selector columns in the base cost.

I looked into the `y_switch` constraint, and I don't see any way to remove its `prev` query or alter the column it is in (as the relevant region is built around two consecutive complete additions, which already consume most of the cells across 9 columns).
